base/contracts.jl

```diff
@@ -0,0 +1,22 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+
+@noinline function precondition_error(msg)
+    io = IOBuffer()
+    StackTraces.show_spec_linfo(io, StackTraces.lookup(backtrace()[3]))
+    throw(ArgumentError("$(takebuf_string(io)) requires $msg"))
+end
+
+
+"""
+    @require precondition [message]
+
+Throw `ArgumentError` if `precondition` is false.
+"""
+macro require(precondition, msgs...)
+    msg = isempty(msgs) ? string(precondition) : msgs[1]
+    :(if ! $(esc(precondition)) precondition_error($msg) end)
+end
```

Comment:
Since macros can have optional arguments now `@require` could be defined a bit shorter as

```julia
macro require(precondition, msg = string(precondition))
    :(if ! $(esc(precondition)) precondition_error($(esc(msg)) end)
end
```

(+1 to having a `@require` in Base.)